### ISSUE TITLE
[FIRApp configure] takes ~1 second on iPhone 6Plus

### ISSUE THREAD

**Issue by seblau** [firebase/firebase-ios-sdk#2148]
* Xcode version: 10.1
  * Firebase SDK version: 5.11.0
  * Firebase Component: Core, Messaging
  * Component version: 5.11.0

#### Steps to reproduce:

I annotated my AppDelegate.m with two NSLog statements, made a Release build and run the app on an iPhone 6Plus. As you can see from the Log output below, `[FIRApp configure]` takes 0,8 seconds to finish. I testet multiple times (warm and cold starts) and the time ranges between 0,8 and 1,3 second approximately.

#### Questions:
* Can someone reproduce this behaviour or is someone else aware that this takes ~1 second to complete?
* Right now this totally effects the app launch time since its blocking. Is there an options to move `[FIRApp configure]` to some other part of the app startup, so it does not effect app startup and can run somewhere in the background?
* Any ideas why this takes so long?

**AppDelegate.m**
```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions
{
  NSLog(@"didFinishLaunchingWithOptions START");
  [FIRApp configure];
  NSLog(@"didFinishLaunchingWithOptions after [FIRApp configure]");
  [ ... ]
}
```

**Console Logs**
```
===> default	10:44:19.441526 +0100	HelloAgain	didFinishLaunchingWithOptions START
default	10:44:19.517598 +0100	HelloAgain	Task <3B177D5F-2234-4F43-895C-86754B5AE0B9>.<1> resuming, QOS(0x19)
default	10:44:20.207480 +0100	HelloAgain	Faulting in NSHTTPCookieStorage singleton
default	10:44:20.207897 +0100	HelloAgain	Faulting in CFHTTPCookieStorage singleton
default	10:44:20.208161 +0100	HelloAgain	Creating default cookie storage with default identifier
default	10:44:20.208353 +0100	HelloAgain	Task <3B177D5F-2234-4F43-895C-86754B5AE0B9>.<1> {strength 0, tls 4, ct 0, sub 0, sig 1, ciphers 0, bundle 0, builtin 0}
default	10:44:20.216726 +0100	HelloAgain	TIC Enabling TLS [1:0x283b3f9c0]
default	10:44:20.217013 +0100	HelloAgain	TIC TCP Conn Start [1:0x283b3f9c0]
default	10:44:20.218828 +0100	HelloAgain	[C1 Hostname#a6f1ebcf:443 tcp, url: https://play.googleapis.com/log, tls] start
default	10:44:20.221859 +0100	HelloAgain	nw_connection_report_state_with_handler_locked [C1] reporting state preparing
default	10:44:20.229830 +0100	HelloAgain	Task <3B177D5F-2234-4F43-895C-86754B5AE0B9>.<1> setting up Connection 1
===> default	10:44:20.252982 +0100	HelloAgain	didFinishLaunchingWithOptions after [FIRApp configure]
```

**Comment by ryanwilson** [comment]
Hi @seblau - thanks for the report. This doesn't sound great, we'll definitely try to track it down. I'll answer your questions first, then ask a few of my own if that's okay!

> Can someone reproduce this behaviour or is someone else aware that this takes ~1 second to complete?

I'll try to find an iPhone 6 Plus to reproduce - but it would help to know what iOS version it's running (see questions below).

> Right now this totally effects the app launch time since its blocking. Is there an options to move [FIRApp configure] to some other part of the app startup, so it does not effect app startup and can run somewhere in the background?

It's recommended to run `[FIRApp configure]` as soon as possible in order to detect some events early in the app's lifecycle - we don't test running it later, so you may run into issues by delaying it. This delay is definitely not intentional and I consider it a performance bug, so you shouldn't have to move it.

> Any ideas why this takes so long?

I'm not sure off the top of my head, but one guess could be slow keychain reads as we've seen a slow startup issue last year. That's not based on any investigation, though. 🙂


A few questions from me to narrow it down further:

1. What iOS version is the iPhone 6 Plus running?
2. Have you been able to reproduce this on other devices, or only the 6 Plus?
3. If you're using CocoaPods: can you list your `Podfile` (if you're comfortable, but at least the Firebase pods)
4. If you're not using CocoaPods, can you give a list of all Firebase frameworks you're including?
5. Finally, are you able to reproduce this on the most recent version of Firebase? `5.14.0` was released yesterday. This will help us figure out if it's `5.11.0` specific or a larger issue at hand.

Thanks again!

**Comment by davidair** [comment]
Another interesting thing worth checking - what happens if you create a blank app from scratch and add Firebase to it? Would you still encounter the same startup latency issue?

**Comment by seblau** [comment]
Hi @ryanwilson,

1.) iOS 12.0.1
2.) I testet only on iPhone 6Plus so far
3.) For the "warm start" I completely closed the app (swiped it away) and started it again
4.) My test app is a React Native app which includes the Firebase SDK via https://github.com/invertase/react-native-firebase version 5.1.1. This package is using Firebase SDK component of version 5.11.0
5.) The Podfile for the Firebase pods is the following:
```
s.dependency "Firebase/Core", "~> 5.11.0"
s.dependency "Firebase/Messaging", "~> 5.11.0"
```
6.) I will try it with version `5.14.0` as soon as `react-native-firebase` upgrades to this version.

**Comment by morganchen12** [comment]
@seblau can you profile your launch time via Instruments to see which calls are taking up the most time?

**Comment by morganchen12** [comment]
Closing since this issue has staled; please comment and I'll reopen it.

**Comment by ankitgarg1mg** [comment]
I ran profiling on iPhone 6, OS Version 12.2. It took for me also 50ms. I think its too long time. Any suggestions?. I am doing this in didFinishLaunchingOptions. @ryanwilson 

<img width="1033" alt="Screen Shot 2019-04-26 at 6 41 09 PM" src="https://user-images.githubusercontent.com/34055785/56810070-0bbcaf00-6853-11e9-8b39-c2196537efc6.png">

**Comment by ryanwilson** [comment]
@ankitgarg1mg can you please share:

- Which pods you are using
- An expanded view of the profiler image to see which methods are taking the longest
- What an acceptable time in your use case would be?

Thanks

**Comment by ankitgarg1mg** [comment]
Hi Ryan,

We are using pod 'FBSDKCoreKit', '4.33.0'. 
<img width="1034" alt="Screen Shot 2019-04-26 at 7 55 59 PM" src="https://user-images.githubusercontent.com/34055785/56814713-5e9b6400-685d-11e9-987c-ac164f702860.png">
Any tips ori suggestion to improve this time. Should I put this code inside concurrent queue with priority as background.

**Comment by ryanwilson** [comment]
@ankitgarg1mg That's the Facebook SDK pod, can you share which Firebase specific ones you're using? They should all be prefixed with "Firebase". I can see at least Analytics, Remote Config, and Dynamic links from there.

**Comment by ankitgarg1mg** [comment]
pod 'Firebase/Core', '5.20.2'

**Comment by morganchen12** [comment]
@ankitgarg1mg 57ms is roughly three frames of processing time. It's unlikely we'll be able to make `-[FIRApp configure]` significantly faster than that, but we may be able go find some small time savings. You **should not** put the configuration call in a low-priority background thread, since [low priority threads are not guaranteed to make progress on iOS](https://lists.swift.org/pipermail/swift-dev/Week-of-Mon-20151214/000372.html).

**Comment by ankitgarg1mg** [comment]
@ryanwilson I am optimizing my app cold start time. I was thinking to optimize didFinishLaunching method too. Should I put this code inside a concurrent queue with priority as high? I do not want to initialize this on the main thread. I am using FIR for remote notification only. Hope you understand my problem

**Comment by RealHacker** [comment]
When the firebase server is not reachable (network is ok, but firebase servers blocked by firewall), FIRApp.configure takes 19 seconds to retry connecting. My app shows the splash screen for that long...

Is it OK to run FIRApp.configure in another thread?

**Comment by morganchen12** [comment]
@RealHacker is this reproducible on devices with airplane mode on? You should not configure Firebase in a background thread.

**Comment by RealHacker** [comment]
You only need to be in China to reproduce this :(. Most Google services are blocked here.
I guess you can set up a proxy server and change hosts to reproduce this.

Extra info:
This only happens for the first time launch of the app.
- Firebase, 3.17.0
- FirebaseAnalytics, 3.9.0
- FirebaseCore, 3.6.0

I used quite an old version, maybe this has been fixed. I will update my pods and post results.

**Comment by RealHacker** [comment]
Please ignore my comments above. I cannot reproduce the delay with the latest version.
`Time for firebase initialization: 0.11451101303100586`
